handbook/best_practices.md

```diff
@@ -0,0 +1,137 @@
+11. **Check license compatibility**: Select a project with a license that is compatible with your intended use. Ensure the license aligns with your organization's legal and operational requirements to avoid potential conflicts.
+
+12. **Consider the project's lifecycle stage**: Evaluate where the project is in its lifecycle. Avoid projects in the "hype" stage unless you are willing to take risks or contribute actively. Mature projects with a stable community and fewer breaking changes are often more dependable.
+
+13. **Assess dependency management**: Be cautious of projects with numerous or unstable dependencies. Fewer dependencies often mean less risk of breakage and easier maintenance.
+
+14. **Perform benchmarks**: If you are considering multiple competing solutions, perform internal benchmarks using your specific use cases and data. This ensures the chosen software meets your performance and functionality requirements.
+
+
+## Creating secure containers
+
+Creating secure containers is essential to minimize vulnerabilities and ensure the reliability of your applications. Below are some best practices to follow when building container images:
+
+### Use multi-stage builds
+
+Multi-stage builds help reduce the size of the final container image by separating the build environment from the runtime environment. This approach ensures that only the necessary artifacts are included in the final image, excluding build tools and dependencies.
+
+For example, when building a Go application:
+```Dockerfile
+# Stage 1: Build
+FROM golang:1.20 AS builder
+WORKDIR /app
+COPY . .
+RUN go build -o myapp
+
+# Stage 2: Runtime
+FROM alpine:3.18
+WORKDIR /app
+COPY --from=builder /app/myapp .
+CMD ["./myapp"]
+```
+
+### Use minimal base images
+
+Whenever possible, use minimal base images like `Alpine` to reduce the attack surface and the number of vulnerabilities. For example:
+
+```Dockerfile
+FROM alpine:3.18
+```
+However, be cautious when using minimal images, as some software may not be fully compatible with them. For instance, MUSL (used in Alpine) can cause issues with certain applications. In such cases, consider using a more compatible base image like `Debian`,
+maybe with the `slim` variant:
+
+```Dockerfile
+FROM debian:12.11-slim
+```
+
+### Pin image versions
+
+Always specify the version of the base image to ensure consistency and avoid unexpected changes. For example:
+```Dockerfile
+FROM alpine:3.18
+```
+Avoid using untagged or `latest` tags, as they can lead to unpredictable builds and runtime issues.
+
+### Consider dependency pinning
+
+When installing packages with managers like `apk` or `apt`, you may choose to pin package versions to improve build reproducibility. For example:
+```Dockerfile
+RUN apk add --no-cache openssl=3.0.9-r1
+```
+However, strict pinning can delay important security updates. Weigh the benefits of reproducibility against the need for timely patches, and choose an approach that fits your project's requirements.
+
```

Comment:
I'd add an alternative to dependency pinning, the [`SOURCE_DATE_EPOCH`](https://reproducible-builds.org/docs/source-date-epoch/) is now a standardized variable throughout various environments, that even base images for containers support.

This will make build fail if you update to a base image that cannot have the dependencies that respect your timestamp, giving immediate feedback on something to check.

Reply:
We use it in nethsec!

https://github.com/NethServer/nethsecurity/blob/b581cd0d06f2cb5553ee663d4c6fdb0b032feba4/builder/Containerfile#L8